Updating Zydis-Pascal to the latest Zydis version

Coldzer0 opened the pull request:
This work is still in progress.

I'm converting the code based on the examples ^_^
So, all structs and APIs converted are the ones that are needed for the examples to work.

I'll update this based on examples till I convert all of them, and then I will add any other needed APIs.

flobernd commented on README.md:
```diff
@@ -1,90 +1,68 @@
+## TODO
+This repo still in development & Supports only FPC for now
 
-#### Dynamic
-- Extract the `Zydis32.dll` or `Zydis64.dll` from the `Bin32`/`Bin64` folder to the directory that contains your application or install the library to your `C:\Windows\System32\` (64-bit) / `C:\Windows\SysWOW64\` (32-bit) directory
-- Enable the `{$DEFINE ZYDIS_DYNAMIC_LINK}` compiler-directive (enabled by default) in `Zydis.pas`
-
-#### Static
-- Extract the complete `Bin32`/`Bin64` folder to the root of the pascal bindings directory
-- Disable the `{$DEFINE ZYDIS_DYNAMIC_LINK}` compiler-directive in `Zydis.pas`
```
I'm not a super huge fan of including binary files in a Git repository, but we can consider going that route, if we feel like this would lower the burden of using the bindings.

Anyways, I think that part of the readme should stay. At least the part about dynamic linking might be important to some users - regardless of whether we bundle the static binaries in the repo or not.

flobernd replied:
I think Delphi is still mainly used on Windows, but I'm not sure how the situation is for FPC. It easily gets compilcated when we have to bundle a set of X86(_64) and ARM(64) binaries for different OSs like Windows, Linux, FreeBSD, etc. This was the original reason for not including the binaries in the repo.

flobernd replied:
I think we could create a GitHub Actions release workflow that uses GCC or Clang to cross-compile Zydis for different target architectures and operating systems in different formats (`.a`, `.obj`, `.dll`, ...).

Coldzer0 replied:
I agree that `dynamic linking` is essential for many users.
For me, I only use Delphi for User Interface projects, but Pascal, I do use it for all platforms.

GitHub Actions release workflow would be great for auto-bundling Zydis binaries.

flobernd commented on README.md:
```diff
@@ -1,90 +1,68 @@
-## Readme
-The Zydis pascal bindings support both dynamic and static linkage. The required binaries for Windows can be acquired in the [Releases](https://github.com/zyantific/zydis-pascal/releases) section.
+## TODO
+This repo still in development & Supports only FPC for now
```
I'm happy to provide the Delphi projects. We should consider going with `*.dpr`, `*.dfm`, .etc as AFAIK they can be opened with FPC/Lazarus as well but the other way is not possible.

Coldzer0 replied:
I do have Delphi, but I'll do most of the work on FPC first and then test it with Delphi.
I'll change the `.lpr` to `.dpr`

flobernd commented on Zydis/Zydis.Decoder.pas:
These files were originally ment to provide a high level wrapper around the native APIs so that Zydis can be used in a more natual way (aka using classes).

We might want to check how files should be organized. I like the way you split the APIs from just `Zydis.pas` to sub-units, but maybe we could do something like `./Zydis/API/Zydis.Decoder.pas`, `./Zydis/API/Zydis.Formatter.pas` etc. for the native APIs and keep everything in the root for the high level wrapper like this file `./Zydis/Zydis.Decoder.pas`.

Coldzer0 replied:
@flobernd 

This will create a conflict in naming for Pascal users.

What I've in mind is to put all APIs in `Zydis.Apis.pas` and the types in their name convention like `Zydis.Decoder.Types` and `Zydis.Disassembler.Types` etc.

This way, we can Have the `Zydis.Disassembler` and `Zydis.Decoder` and other units for the high-level code.
So, if someone wants to use the low-level code, they can use the `Zydis.Apis.pas` and the `Types` units, or for high-level, they can use the one without it.

flobernd commented on Zydis/Zydis.apis.pas:
nit: It would be great, if we could use PascalCased filenames 🙂

Coldzer0 replied:
On it 👍

flobernd reviewed (commented):
Thanks a lot for your work! I've not been using Delphi/Pascal for a while and definitely would not have been able to design the updated bindings in such a nice and compatible way.

I left a few (rather minor) comments and suggestions 🙂

Coldzer0 commented:
> Thanks a lot for your work! I've not been using Delphi/Pascal for a while and definitely would not have been able to design the updated bindings in such a nice and compatible way.
> 
> I left a few (rather minor) comments and suggestions 🙂

You are welcome ^_^, And thanks for your kind words.

Coldzer0 commented:
@flobernd 

I added `EncodeFromScratch` example and needed Types and APIs.
I tested all three examples on Windows & Linux - x64.

I moved all Types from `Disassembler`, `Decoder`, and `Formatter` Units to new Units; 
[5291745](https://github.com/zyantific/zydis-pascal/pull/2/commits/5291745fc0a4baa58cfa7d07202c29c910d09601)
[49adc6f](https://github.com/zyantific/zydis-pascal/pull/2/commits/49adc6f7ae93cc91a6dd5b1a06730861d2f0c20b)
[447f253](https://github.com/zyantific/zydis-pascal/pull/2/commits/447f253387d48faeef96e589e0743d4beb0a6c7c)

this way, we can use them for high-level implementation.

Coldzer0 commented:
Now, all examples are Delphi compatible.
And compiles for Linux with FPC.

flobernd commented on Zydis/Zydis.Disassembler.pas:
I think this should be the Zydis.Encoder, right? The „disassembler“ is equivalent to the decoder, but for the encoder we don’t have a high level unit yet.

Coldzer0 replied:
Yes, that's true.
I've updated it.

flobernd commented on Zydis/Zydis.Enums.pas:
```diff
@@ -0,0 +1,69 @@
+{
+  Defines the `TZyanMemoryPageProtection` enum.
+}
+type
```
This technically is a Zycore enum, but I guess having it here should be fine for now 🙂

Coldzer0 replied:
Yep, I didn't notice ^_^
I'll change it in the next pull request.

flobernd reviewed (commented):
Thanks again for all the work! This looks really good 🙂 Let me know when you feel like you are done for now. We can merge the PR and improve/complete it in some follow up PRs.

I’ll try to find some time to work on the GitHub Action to generate the binary files.

Coldzer0 commented:
@flobernd 

You can merge this pull request, and I'll open a new one for the other examples and updates.
